### ISSUE TITLE
make deps broken for MacOS builds

### ISSUE THREAD

**Issue by jakubgs** [status-im/status-desktop#2255]
It appears that the #2244 has introduced this error in CI for MacOS builds:
```
05:23:27  + make deps
05:23:27  \e[92mFetching:\e[39m bottles for macOS
05:23:27  \e[92mBuilding:\e[39m Nim compiler
05:23:30    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
05:23:30                                   Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
05:23:30  
05:23:30  tar: Error opening archive: Unrecognized archive format
05:23:30  make: *** [bottles/pcre/INSTALL_RECEIPT.json] Error 1
05:23:30  make: *** Waiting for unfinished jobs....
```

**Comment by jakubgs** [comment]
This appears to be culprit:
https://github.com/status-im/status-desktop/blob/63bc6f0e536ab2d6310e9f45f9be8ae510eb62bd/Makefile#L114-L123

**Comment by jakubgs** [comment]
It appears the file we get is not a GZip archive:
```
jenkins@macos-01.ms-eu-dublin.ci.misc:~ % URL=$(brew info --json=v1 pcre | jq -r '.[0].bottle.stable.files.mojave.url')

jenkins@macos-01.ms-eu-dublin.ci.misc:~ % echo $URL
https://ghcr.io/v2/homebrew/core/pcre/blobs/sha256:ed9b483538da7bc6559d2e63dd36659736fab9510681661d970d707a18731de4

jenkins@macos-01.ms-eu-dublin.ci.misc:~ % curl -L -o openssl.tar.gz -u _:_ $URL
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                     
                                 Dload  Upload   Total   Spent    Left  Speed
100    83  100    83    0     0    351      0 --:--:-- --:--:-- --:--:--   351
jenkins@macos-01.ms-eu-dublin.ci.misc:~ % file openssl.tar.gz 
openssl.tar.gz: JSON data
```

**Comment by jakubgs** [comment]
Ah, there we have it:
```
jenkins@macos-01.ms-eu-dublin.ci.misc:~ % cat openssl.tar.gz
{"errors":[{"code":"UNAUTHORIZED","message":"authentication token not provided"}]}
```

**Comment by jakubgs** [comment]
And indeed, we do get a `401`:
```
jenkins@macos-01.ms-eu-dublin.ci.misc:~ % curl -isL -u _:_ $URL
HTTP/2 401 
content-type: application/json
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://ghcr.io/token",service="ghcr.io",scope="repository:homebrew/core/pcre:pull"
date: Thu, 15 Apr 2021 10:25:21 GMT
content-length: 83
x-github-request-id: D441:72ED:2C1E05:3568C6:60781491

{"errors":[{"code":"UNAUTHORIZED","message":"authentication token not provided"}]}
```

**Comment by jakubgs** [comment]
According to their docs there is an env variable that is supposed to provide the token:

>`HOMEBREW_GITHUB_PACKAGES_TOKEN`
>Use this GitHub personal access token when accessing the GitHub Packages Registry (where bottles may be stored).

https://docs.brew.sh/Manpage

But where that is supposed to come from? No idea. Not even sure why we are downloading it like this.

**Comment by jakubgs** [comment]
Also, it seems to me like this is not exactly the safest way of getting something like OpenSSL.

You're not even checking a checksum for what you download, nor is the version pinned.

**Comment by jakubgs** [comment]
Seems like `ghcr.io` is the default in Homebrew:
https://github.com/Homebrew/brew/blob/3.1.2/Library/Homebrew/env_config.rb#L47-L58

**Comment by jakubgs** [comment]
According to @michaelsbradleyjr the requests fail because they are throttled when no authentication is provided, so that explains why it initially worked, and then started failing. Will definitely can't rely on that to work every time.

**Comment by jakubgs** [comment]
I'd propose downloading the bottles and then uploading them to our own DigitalOcean bucket, and then pin it in the repo and verify SHA256 on download. Apparently OpenSSL doesn't update that often, so we could manage that by hand.

**Comment by michaelsbradleyjr** [comment]
> Ah, there we have it:
> 
> ```
> jenkins@macos-01.ms-eu-dublin.ci.misc:~ % cat openssl.tar.gz
> {"errors":[{"code":"UNAUTHORIZED","message":"authentication token not provided"}]}
> ```

In the [PR](https://github.com/status-im/status-desktop/pull/2244) description that changed where the bottles are downloaded from I wrote:

> At this time downloading from GitHub Packages requires specifying a username and personal acccess token, but it seems they don't have to be valid. For example, in the changes to the Makefile I specified `_:_` as the username and token.

So I guess GitHub must be rate limiting how many times you can hit that endpoint without a valid username:token pair.

> According to their docs there is an env variable that is supposed to provide the token:
> 
> > `HOMEBREW_GITHUB_PACKAGES_TOKEN`
> > Use this GitHub personal access token when accessing the GitHub Packages Registry (where bottles may be stored).
> 
> https://docs.brew.sh/Manpage
> 
> But where that is supposed to come from? No idea. Not even sure why we are downloading it like this.

That's a token they're embedding somewhere in the builds of Homebrew. For local development builds, I'll change the Makefile so that curl uses `-u $(GH_USER):$(GH_TOKEN)`. Each developer building for macOS will need to get a token from their account on github.com and setup the environment variables. We'll handle CI as you suggest (see below).

> Also, it seems to me like this is not exactly the safest way of getting something like OpenSSL.
> 
> You're not even checking a checksum for what you download, nor is the version pinned.

We weren't checking the hash any time since the start of the desktop project; the only thing that changed in my recent PR was that we were downloading from bintray, and now we're downloading from GitHub Packages.

However, it's good practice to check it so I'll make a change to the Makefile to check it after download. The sha256 to check against is available in the output of e.g. `brew info --json=v1 openssl`.

As for pinning, OpenSSL releases security fixes a few times per year, and by downloading whatever is the latest Homebrew build for the oldest macOS for which Homebrew builds bottles, then we automatically pick up those fixes.

However, if we want to pin it in CI, we can. I'll jiggle the Makefile variables for `BOTTLE_OPENSSL` and `BOTTLE_PCRE` so that they can be overriden in the environment. Then in Jenkins CI we can invoke `make` with those set to a paths where they've been downloaded from "our own DigitalOcean bucket". Does that sound like a decent plan?

**Comment by jakubgs** [comment]
Yeah, I guess normal devs won't hit the throttling because they would download it just once, while our CI hosts download them multiple times a day.

I'm not happy to add edge cases to CI, but it's better than having it fail randomly. But then again, what's the benefit if you build releases in CI? Then you have to make sure you keep the OpenSSL/PCRE version updated in CI before releases, how is that different from keeping it updated in the repo?

**Comment by michaelsbradleyjr** [comment]
> I'm not happy to add edge cases to CI, but it's better than having it fail randomly. But then again, what's the benefit if you build releases in CI? Then you have to make sure you keep the OpenSSL/PCRE version updated in CI before releases, how is that different from keeping it updated in the repo?

Before the changes in #2244 the exact bottles we were downloading were more hardwired than they should have been, I just hadn't thought about it in a long time:

```
https://bintray.com/homebrew/bottles/download_file?file_path=openssl%401.1-1.1.1g.high_sierra.bottle.tar.gz
https://bintray.com/homebrew/bottles/download_file?file_path=pcre-8.44.high_sierra.bottle.tar.gz
```

We were using those the whole time even though there have been newer versions of the openssl bottle published.

With the changes in #2244 it's no longer hardwired, i.e. the latest version of the openssl and pcre bottles built for Mojave (10.14) will be downloaded. 

So an option would be in CI to setup `GH_USER` and `GH_TOKEN` in the environment and then CI will download whatever is the latest and we don't have to keep the versions updated in the repo nor in CI. I think that approach would be the simplest, but not sure how you feel about it.

I still need to create the PR that uses `GH_USER` and `GH_TOKEN`, but I'll have that done later today.

Side note: some day in the future Homebrew will stop building bottles for Mojave and I'll need to change it to Catalina (10.15). It's hard to predict when that will happen, but we'll know when it does because the CI builds will break.

**Comment by michaelsbradleyjr** [comment]
See #2258.

**Comment by michaelsbradleyjr** [comment]
Closed by #2258.
